collect_app/src/test/java/org/odk/collect/android/http/CollectServerClientTest.java

```diff
@@ -31,9 +37,20 @@ public void setup() {
 
     @Test
     public void testGetXMLDocumentErrorResponse() {
-        collectServerClient.setGetHttpShouldReturnNull(true);
-        DocumentFetchResult fetchResult = collectServerClient.getXmlDocument("http://testurl");
-        assertEquals(fetchResult.errorMessage, "Parsing failed with null while accessing http://testurl");
+        final String urlString = "http://testurl";
+
+        MockHttpClientConnection clientConnection = mock(MockHttpClientConnection.class);
+
+        try {
+            URL url = new URL(urlString);
+            doReturn(null).when(clientConnection).get(url.toURI(), CollectServerClient.HTTP_CONTENT_TYPE_TEXT_XML, null);
```

Comment:
Awesome John. Glad to see that this idea actually works!
I hope this isn't too much, but I would even go a step further to make both `CollectServerClient` instances injectable. What we are doing here is really important as we are setting the groundwork for other tests to utilize these ideas to get the most out of testing these use cases with minimal effort. For eg. a way to clean this up would be to utilize the `Named` annotation so you would be able to provide two different server clients. One that returns null when you instantiate it, so we could call that one `Named("NullGet")` and the other works normally. So in this case, one client would be using the Mockito version of the Http client while the other would utilize the hard wired version. This is just an idea and probably not necessary but just putting thoughts out there. You could probably move this code here into the `setup` method since it's more along the lines of configuring this client for testing.

Reply:
Hi @jd-alexander, when you say make `CollectServerClient` injectable, do you mean throughout the app or for testing purposes?